triton-vm/src/table/hash_table.rs

```diff
@@ -765,36 +781,43 @@ impl ExtHashTable {
             Self::mode_deselector(circuit_builder, &mode, HashTableMode::Pad)
                 * Self::select_mode(circuit_builder, &mode_next, HashTableMode::Pad);
 
-        let difference_of_capacity_registers = state_current[RATE..]
-            .iter()
-            .zip_eq(state_next[RATE..].iter())
-            .map(|(current, next)| next.clone() - current.clone())
-            .collect_vec();
-        let randomized_sum_of_capacity_differences = state_weights[RATE..]
-            .iter()
-            .zip_eq(difference_of_capacity_registers)
-            .map(|(weight, state_difference)| weight.clone() * state_difference)
-            .sum::<ConstraintCircuitMonad<_>>();
-
-        let capacity_doesnt_change_at_section_start_when_program_hashing_or_absorbing =
+        let randomized_sum_of_capacity_differences = izip!(
+            &state_weights[RATE..],
+            &state_current[RATE..],
+            &state_next[RATE..]
+        )
+        .map(|(weight, current, next)| weight.clone() * (next.clone() - current.clone()))
+        .sum::<ConstraintCircuitMonad<_>>();
+
+        // If the next round number is 0, then
+        // - the next mode is either `Hash` or `Pad`, or
+        // - the next instruction is `sponge_init`, or
+        // - the capacity does not change, or
+        // - the current mode is `ProgramHashing` and (!) the next mode is `sponge`.
+        //
+        // The “and” marked with an exclamation mark requires above constraint to be
+        // expressed through two polynomials.
+        let capacity_doesnt_change_at_section_start_when_program_hashing_or_ =
```

Comment:
I think `section_start` is a poor choice of words. What is a section? Are there at most four sections, one for each mode? I gather (perhaps incorrectly) that `section` means the same as `round`. Why not use that word instead? Or even better: `new_round`.

Reply:
If it does mean "contiguous region of constant mode" I would suggest using instead `mode_switch` instead.

Reply:
This (pair of) constraints is very confusing. Is it because it is trying to achieve several features at once? If so, consider separating it into more constraints. If mode switches from A to B then enforce C; if mode switches from X to Y then enforce Z; etc.